# Exercise 1: Collinearity (by hand)
## (a)
The sample mean of $Y$ is $\bar{Y}_N = (0 + 3 + 6)/3 = 3$ and for $X_1$, $\bar{X}_N = (3 + 4 + 5)/3 = 4$. Therefore, the numerator is
$$
(0 - 3)(3 - 4) + (3 - 3)(4 - 4) + (6 - 3)(5 - 4) = 3 + 0 + 3 = 6
$$
and the denominator is
$$
(3 - 4)^2 + (4 - 4)^2 + (5 - 4)^2 = 1 + 0 + 1 = 2,
$$
so that $\hat{\beta}_1 = 6/2 = 3$. Then $\hat{\beta}_0 = 3 - (3)4 = -9$.

## (b)
First lets obtain $\boldsymbol{X}'\boldsymbol{X}$, which is
$$
\begin{pmatrix}
1 & 1 & 1 \\ 3 & 4 & 5
\end{pmatrix}
\begin{pmatrix}
1 & 3 \\ 1 & 4 \\ 1 & 5
\end{pmatrix} =
\begin{pmatrix}
1\cdot 1 + 1\cdot 1 + 1\cdot 1 & 1\cdot 3 + 1\cdot 4 + 1\cdot 5 \\
3\cdot 1 + 4\cdot 1 + 5\cdot 1 & 3\cdot 3 + 4\cdot 4 + 5\cdot 5
\end{pmatrix} = 
\begin{pmatrix}
3 & 12 \\
12 & 50
\end{pmatrix}.
$$
Next, we take the inverse of this matrix using the formula for the inverse of a 2x2 matrix
$$
A = \begin{pmatrix}
a & b \\
c & d
\end{pmatrix}
$$
where
$$
A^{-1} = \frac{1}{ad - bc}\begin{pmatrix}
d & -b \\
-c & a
\end{pmatrix},
$$
so that in our example
$$
(\boldsymbol{X}'\boldsymbol{X})^{-1} = 
\frac{1}{150 - 144}\begin{pmatrix}
50 & -12 \\
-12 & 3
\end{pmatrix} =
\begin{pmatrix}
\frac{25}{3} & -2 \\
-2 & \frac{1}{2}.
\end{pmatrix}
$$
Next we determine $\boldsymbol{X}'\boldsymbol{Y}$.
$$
\boldsymbol{X}'\boldsymbol{Y} = \begin{pmatrix}
1 & 1 & 1 \\ 3 & 4 & 5
\end{pmatrix}
\begin{pmatrix}
0 \\ 3 \\ 6
\end{pmatrix} = 
\begin{pmatrix}
1\cdot 0 + 1\cdot 3 + 6\cdot 1 \\
3\cdot 0 + 3\cdot 4 + 5\cdot 6
\end{pmatrix} = 
\begin{pmatrix}
9 \\ 42
\end{pmatrix}.
$$
Therefore,
$$
(\boldsymbol{X}'\boldsymbol{X})^{-1}\boldsymbol{X}'\boldsymbol{Y} = 
\begin{pmatrix}
\frac{25}{3} & -2 \\
-2 & \frac{1}{2}
\end{pmatrix}
\begin{pmatrix}
9 \\ 42
\end{pmatrix} = 
\begin{pmatrix}
-9 \\ 3
\end{pmatrix},
$$
and we obtain estimates identical to part (a). If you want, you can confirm the result in R.

In [12]:
X = cbind(rep(1, 3), c(3, 4, 5))
Y = c(0, 3, 6)
b = solve(t(X)%*%X)%*%t(X)%*%Y
b

[,1]
[1,] -9  
[2,]  3

## (c)
We need to assume that $E[XX']$ is invertible, or equivalently, that there is no perfect collinearity in $X$. This means that $\forall$ non-zero $c\in\boldsymbol{R}^2$, $P(X'c=0)<1$. In the $2\times 2$ case, 
$$
E[XX'] = \begin{pmatrix}
1 & E[X_1] \\
E[X_1] & E[X_1^2]
\end{pmatrix}.
$$
A matrix is invertible if and only if it has a non-zero determinant. The determinant of $E[XX']$ is $E[X_1^2] - E[X_1]^2 = Var[X_1]$. If this is non-zero, and since $Var[X_1]\geq 0$, then $Var[X_1]>0$. 

## (d)
If the sample variance is really small the magnitudes of $\hat{\beta}_1$ and $\hat{\beta}_0$ are large. The sample counterpart to the conditions in (c) are that $\boldsymbol{X}'\boldsymbol{X}$ is invertible, or more simply, in this univariate cases, that the sample variance of $X_1$ is greater than 0. Thus, if there is ANY variance in our data, we can estimate $\beta$. The idea about the magnitude of $\hat{\beta}_1$ and $\hat{\beta}_0$ exploding is that we essentially run into an identification problem. If there actually was no variation in $X_1$, i.e., $Var[X_1]=0$, then we are essentially trying to describe the mean of Y, $E[Y]$ using two constants, $\beta_0$ and $\beta_1$. But since there are an infinite number of possible $\beta_0$ and $\beta_1$ such that $\beta_0 + \beta_1 = E[Y]$, this can be reflected in the variance of our sample estimators $\hat{\beta}_0$ and $\hat{\beta}_1$ if the sample variance of $X_1$ is small.

## (e)
It is obvious that $\boldsymbol{X}$ is rank 2 because you cannot write $(1,1,1)'$ as a linear combination of $(3,4,5)'$. Suppose $\exists c\neq 0$ such that $c(1,1,1)'= (3,4,5)'$. Then $c=3$ and $c=4$, a contradiction. From part (b), we have that
$$
\boldsymbol{X}'\boldsymbol{X} = \begin{pmatrix}
3 & 12 \\
12 & 50
\end{pmatrix}.
$$
One could use the same proof, on both the columns and rows, to show that there does not exist $c\neq 0$, such that $c(3, 12)'= (12, 15)'$, so that the rows and columns are linearly independent, and therefore this matrix is of rank 2. The determinant is just $3\cdot 50 - 12\cdot 12 = 150 - 144 = 6\neq 0$. To find the eigenvalues remember that an eigenvalue is defined as a $\lambda$ such that $\boldsymbol{X}'\boldsymbol{X}v = \lambda v$ for some non-zero $v$. This implies that $(I\lambda - \boldsymbol{X}'\boldsymbol{X})v=0$, so that by solving for the $\lambda$ in the expression $det(I\lambda - \boldsymbol{X}'\boldsymbol{X})=0$, where $det$ denotes determinant, we can find our eigenvalues $\lambda$. Writing this out we have
$$
det\begin{pmatrix}
\lambda - 3 & -12 \\ -12 & \lambda - 50
\end{pmatrix} = (\lambda - 3)(\lambda - 50) - 144 = 0
$$
so that $\lambda^2 - 53\lambda + 150 - 144 = 0$ for which we obtain solutions
$$
\lambda = \frac{53 \pm \sqrt{53^2 - 24}}{2},
$$
which we can find on our computer:

In [16]:
c((53 + sqrt(53^2 - 24))/2, (53 - sqrt(53^2 - 24))/2)

[1] 52.8865496  0.1134504

Which we can confirm in R.

In [17]:
eigen(t(X)%*%X)$values

[1] 52.8865496  0.1134504

We can also confirm our result for the determinant.

In [13]:
det(t(X)%*%X)

[1] 6

Knowing that $\boldsymbol{X}'\boldsymbol{X}$ is invertible, or equivalently that the columns of $\boldsymbol{X}$ are linearly independent means that each column has it's own information content separate from the other columns. As an example, we saw in class if we had $X=(1,X_1,X_2)'$ and $X_1$ was a male indicator and $X_2$ a female indicator, then $X$ will have the problem of perfect collinearity. This will translate into $\boldsymbol{X}$ having columns that are linearly independent. And the reason is simple: knowing that someone is male is enough to know that they are not female, so using $X_2$ in addition to $X_1$ brings no more informational content.

# Exercise 2: Collinearity (by Computer)

In [27]:
N = 100
set.seed(210)

X1 = rnorm(N, mean=1, sd=15)
X2 = X1 + rnorm(N, mean=0, sd=1)
X = cbind(rep(1, N), X1, X2)
norm.X = apply(X, 2, function(x){sqrt(sum(x^2))})
X.n = sweep(X, 2, norm.X, "/")
X.eig = eigen(t(X.n)%*%X.n)$values
cond.no = sqrt(max(X.eig)/min(X.eig))
cond.no

[1] 31.35198

We see that $\gamma \approx 31.4 > 20$, so our data exhibits multicollinearity.

# Exercise 3: Theory and Simulation for Regression

In [11]:
set.seed(210)
N = 1000 # sample size

# Regression parameters
b0 = -3
b1 = 2

U = rnorm(N, mean=0, sd=3)
X1 = rnorm(N, mean=2, sd=2)

Y = b0 + X1*b1 + U

# (a)
Let $X=(1,X_1)'$. In order to estimate $\beta=(\beta_0,\beta_1)'$ by OLS, we must have that $E[XX']$ is invertible or equivalently, that $X$ has no perfect collinearity. Because $X$ contains $X_1$ and a constant, this is equivalent to $Var[X_1]>0$.

In [12]:
b1.hat = cov(Y,X1)/var(X1)
b0.hat = mean(Y) - mean(X1)*b1.hat
c(b0.hat, b1.hat)

[1] -3.039162  1.973260

We obtain estimates very close to the truth.

## (b)
The ith row represents the ith observation $X_i=(1,X_{1,i})'$. The OLS estimator in matrix notation is

$$
\hat{\beta}_N = (\boldsymbol{X}'\boldsymbol{X})^{-1}\boldsymbol{X}'\boldsymbol{Y},
$$

where $\boldsymbol{X}$ is an $N\times 2$ matrix and $\boldsymbol{Y}$ is a $N\times 1$ vector.

In [13]:
X = cbind(rep(1, N), X1)
b.hat = solve(t(X)%*%X)%*%t(X)%*%Y
c(b.hat)

[1] -3.039162  1.973260

And we obtain estimates identical to those in part (a).

# Exercise 4: A Regression of Your Choice